Add Euler polynomials

cbm755 opened the pull request:
`euler(n)` gave Euler numbers.  Extend this to `euler(n, x)` to compute Euler polynomials.  This interface is similar to how `bernoulli` already behaves.

cbm755 committed:
euler numbers: errors for invalid input

Raise errors for Numbers that are something other than nonnegative
integers; this prevents `euler(-3)` from being 0 and `euler(-2)`
from being unevaluated.  Similarly, it is not enough for a Symbol
to be odd, it must also be positive, for `euler(m)` to be 0.

cbm755 committed:
implement euler polynomials

Previously only euler numbers were supported.  Update docs.

cbm755 committed:
euler: fixes for pretty printing and latex

Support Euler polynomials.  Fix raising to power in latex.

cbm755 committed:
printing: Octave support for euler num/poly

Matlab implements these in the Symbolic Toolbox.  Octave will support
them in the Symbolic package.

cbm755 committed:
euler polynomial is rewriteable as sum

cbm755 committed:
euler poly: use mpmath for numerical evaluation

This is faster for large m and (probably) uses a numerically stable
algorithm (instead of evaluating the monomial basis).

cbm755 committed:
euler: fix broken doctest by importing "S"

cbm755 committed:
euler: use LaTeX instead of ascii-art for docs

Should also fix a sphinx-related CI failure.

cbm755 committed:
euler: use result from another CAS for Float test

jksuom commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -736,36 +770,72 @@ class euler(Function):
+                    return Integer(res)
+                # Euler polynomial: numerical evaluation
+                elif sym.is_Float or (sym.is_number and sym.is_complex \
+                        and all([a.is_Float for a in sym.as_real_imag()])):
```
Would it suffice to use `pure_complex` here?

cbm755 replied:
yes!  and doing so, we can improve the logic.  Thanks for showing me `pure_complex`.

cbm755 committed:
euler: rewrite the Float evaluation bit using pure_complex

Thanks @jksuom for pointing this out.  This also simplifies the code
so we choose to numerically evaluate `sym = a + b*I` using the logic
that one of a or b must be a Float and the other could be either a
Float or an Integer.

jksuom commented:
Thanks, this looks good.
